src/alias_asref.rs

```diff
@@ -0,0 +1,310 @@
+    ///
+    /// In particular, if an axis is sliced with an index, the axis is
+    /// collapsed, as in [`.collapse_axis()`], rather than removed, as in
+    /// [`.slice_move()`] or [`.index_axis_move()`].
+    ///
+    /// [`.collapse_axis()`]: Self::collapse_axis
+    /// [`.slice_move()`]: Self::slice_move
+    /// [`.index_axis_move()`]: Self::index_axis_move
+    ///
+    /// See [*Slicing*](#slicing) for full documentation.
+    /// See also [`s!`], [`SliceArg`], and [`SliceInfo`](crate::SliceInfo).
+    ///
+    /// **Panics** in the following cases:
+    ///
+    /// - if an index is out of bounds
+    /// - if a step size is zero
+    /// - if [`NewAxis`](`crate::SliceInfoElem::NewAxis`) is in `info`, e.g. if `NewAxis` was
+    ///   used in the [`s!`] macro
+    /// - if `D` is `IxDyn` and `info` does not match the number of array axes
+    #[track_caller]
+    pub fn slice_collapse<I>(&mut self, info: I)
+    where I: SliceArg<D>
+    {
+        self.as_mut().slice_collapse(info);
+    }
+
+    /// Slice the array in place along the specified axis.
+    ///
+    /// **Panics** if an index is out of bounds or step size is zero.<br>
+    /// **Panics** if `axis` is out of bounds.
+    #[track_caller]
+    pub fn slice_axis_inplace(&mut self, axis: Axis, indices: Slice)
+    {
+        self.as_mut().slice_axis_inplace(axis, indices);
+    }
+
+    /// Slice the array in place, with a closure specifying the slice for each
+    /// axis.
+    ///
+    /// This is especially useful for code which is generic over the
+    /// dimensionality of the array.
+    ///
+    /// **Panics** if an index is out of bounds or step size is zero.
+    #[track_caller]
+    pub fn slice_each_axis_inplace<F>(&mut self, f: F)
+    where F: FnMut(AxisDescription) -> Slice
+    {
+        self.as_mut().slice_each_axis_inplace(f);
+    }
+
+    /// Selects `index` along the axis, collapsing the axis into length one.
+    ///
+    /// **Panics** if `axis` or `index` is out of bounds.
+    #[track_caller]
+    pub fn collapse_axis(&mut self, axis: Axis, index: usize)
+    {
+        self.as_mut().collapse_axis(axis, index);
+    }
+
+    /// Return `true` if the array data is laid out in contiguous “C order” in
+    /// memory (where the last index is the most rapidly varying).
+    ///
+    /// Return `false` otherwise, i.e. the array is possibly not
+    /// contiguous in memory, it has custom strides, etc.
+    pub fn is_standard_layout(&self) -> bool
+    {
+        <Self as AsRef<LayoutRef<_, _>>>::as_ref(self).is_standard_layout()
+    }
+
+    /// Return true if the array is known to be contiguous.
+    pub(crate) fn is_contiguous(&self) -> bool
+    {
+        <Self as AsRef<LayoutRef<_, _>>>::as_ref(self).is_contiguous()
+    }
+
+    /// Return an iterator over the length and stride of each axis.
+    pub fn axes(&self) -> Axes<'_, D>
+    {
+        <Self as AsRef<LayoutRef<_, _>>>::as_ref(self).axes()
```

Comment:
The `AsRef` / `AsMut` aliasing functions serve two purposes. First off, a function like `axes` that got moved to `LayoutRef` would not longer be directly available on a raw view - i.e., `raw_view.axes()` would be an error - because raw views do not dereference. (They would still exist on any array with `S: Data`, though). So these aliases restore that capability to raw views.

The second is a performance consideration, only relevant for mutating functions on shared arrays. Functions defined on `LayoutRef` and `RawRef` *are* directly available to shared arrays, even without these aliases. However, they do that by going through `&mut ArrayRef`, which un-shares the data. But we don't want that in these cases; `LayoutRef` and `RawRef` don't require unique data. So these aliases shadow these mutating functions and create a method that doesn't un-share data.